langserver/workspace_refs.go

```diff
@@ -0,0 +1,155 @@
+package langserver
+
+import (
+	"context"
+	"fmt"
+	"go/build"
+	"go/token"
+	"log"
+	"sort"
+	"strings"
+	"sync"
+
+	"golang.org/x/tools/go/buildutil"
+
+	opentracing "github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go/ext"
+	"github.com/sourcegraph/jsonrpc2"
+	"github.com/sourcegraph/sourcegraph-go/langserver/internal/refs"
+	"github.com/sourcegraph/sourcegraph-go/pkg/lspext"
+)
+
+func (h *LangHandler) handleWorkspaceReference(ctx context.Context, conn JSONRPC2Conn, req *jsonrpc2.Request, params lspext.WorkspaceReferenceParams) ([]lspext.ReferenceInformation, error) {
+	rootPath := h.FilePath(h.init.RootPath)
+	bctx := h.OverlayBuildContext(ctx, h.defaultBuildContext(), !h.init.NoOSFileSystemAccess)
+
+	var pkgPat string
+	if h.init.RootImportPath == "" {
+		// Go stdlib (empty root import path)
+		pkgPat = "..."
+	} else {
+		// All other Go packages.
+		pkgPat = h.init.RootImportPath + "/..."
+	}
+
+	results := refResultSorter{results: make([]lspext.ReferenceInformation, 0)}
+	var wg sync.WaitGroup
+	pkgs := buildutil.ExpandPatterns(bctx, []string{pkgPat})
+	for pkg := range pkgs {
```

Comment:
for a large amount of packages (eg the kubernetes repo?), does this not mean we start thrashing the CPU/memory since we do unbounded parallelism? If we use this pattern in other places in our lsp server, then nevermind for now.